pkg/backend/httpstate/snapshot.go

```diff
@@ -16,49 +16,140 @@ package httpstate
+		return persister.backend.client.PatchUpdateCheckpoint(
+			persister.context, persister.update, deploymentV3, token)
+	}
+
+	deployment, err := persister.marshalDeployment(deploymentV3)
+	if err != nil {
+		return err
+	}
+
+	differ := persister.deploymentDiffState
+
+	// If there is no baseline to diff against, or diff is predicted to be inefficient, use saveFull.
+	if !differ.ShouldDiff(deployment) {
+		if err := persister.saveFullVerbatim(ctx, differ, deployment, token); err != nil {
+			return err
+		}
+	} else { // Otherwise can use saveDiff.
+		diff, err := differ.Diff(ctx, deployment)
+		if err != nil {
+			return err
+		}
+		if err := persister.saveDiff(ctx, diff, token); err != nil {
+			if differ.strictMode {
+				return err
+			}
+			if logging.V(3) {
+				logging.V(3).Infof("ignoring error saving checkpoint "+
+					"with PatchUpdateCheckpointDelta, falling back to "+
+					"PatchUpdateCheckpoint: %v", err)
+			}
+			if err := persister.saveFullVerbatim(ctx, differ, deployment, token); err != nil {
+				return err
+			}
+		}
+	}
+
+	return persister.deploymentDiffState.Saved(ctx, deployment)
+}
+
+func (persister *cloudSnapshotPersister) saveDiff(ctx context.Context,
+	diff deploymentDiff, token string) error {
+	return persister.backend.client.PatchUpdateCheckpointDelta(
+		persister.context, persister.update,
+		diff.sequenceNumber, diff.checkpointHash, diff.deploymentDelta, token)
+}
+
+func (persister *cloudSnapshotPersister) saveFullVerbatim(ctx context.Context,
+	differ *deploymentDiffState, deployment *apitype.UntypedDeployment, token string) error {
+	untypedDeploymentBytes, err := marshalUntypedDeployment(deployment)
+	if err != nil {
+		return err
+	}
+	return persister.backend.client.PatchUpdateCheckpointVerbatim(
+		persister.context, persister.update, differ.SequenceNumber(),
+		untypedDeploymentBytes, token)
+}
+
+func (persister *cloudSnapshotPersister) marshalDeployment(
+	deployment *apitype.DeploymentV3) (*apitype.UntypedDeployment, error) {
+	raw, err := json.MarshalIndent(deployment, "", "")
+	if err != nil {
+		return nil, fmt.Errorf("serializing deployment to json: %w", err)
+	}
+	return &apitype.UntypedDeployment{Deployment: raw, Version: 3}, nil
 }
 
 var _ backend.SnapshotPersister = (*cloudSnapshotPersister)(nil)
 
 func (cb *cloudBackend) newSnapshotPersister(ctx context.Context, update client.UpdateIdentifier,
-	tokenSource *tokenSource, sm secrets.Manager) *cloudSnapshotPersister {
-	return &cloudSnapshotPersister{
+	tokenSource tokenSourceCapability, sm secrets.Manager) *cloudSnapshotPersister {
+
+	p := &cloudSnapshotPersister{
 		context:     ctx,
 		update:      update,
 		tokenSource: tokenSource,
 		backend:     cb,
 		sm:          sm,
 	}
+
+	if cmdutil.IsTruthy(os.Getenv("PULUMI_OPTIMIZED_CHECKPOINT_PATCH")) {
+		p.deploymentDiffState = newDeploymentDiffState()
+
+		if cmdutil.IsTruthy(os.Getenv("PULUMI_OPTIMIZED_CHECKPOINT_PATCH_STRICT")) {
+			p.deploymentDiffState.strictMode = true
```

Comment:
Will we use this in integration tests to validate?

Reply:
Yeah, I was considering adding a simple integration test once the changes deploy on the service side.